### PR TITLE
Add Non-Uniform Border to Border.

### LINKED ISSUE #12583: Support non-uniform Borders with a non-zero borderRadius
https://github.com/flutter/flutter/blob/9ce3ba314eed296812f2176d7a5bbc66e01ba199/packages/flutter/lib/src/painting/box_border.dart#L462

```dart
new BoxDecoration(
    color: color,
    border: new Border(
      left: new BorderSide(color: color),
      top: new BorderSide(color: color),
      bottom: new BorderSide(color: color),
    ),
    borderRadius: new BorderRadius.only(
      topLeft: new Radius.circular(radius),
      bottomLeft: new Radius.circular(radius),
    ),
)
```
Why can't I do this.

I get this error:

```
══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
A borderRadius can only be given for uniform borders.
'package:flutter/src/painting/box_border.dart': Failed assertion: line 462 pos 12: 'borderRadius ==
null'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
https://github.com/flutter/flutter/issues/new

When the exception was thrown, this was the stack:

The following RenderObject was being processed when the exception was fired:
RenderDecoratedBox#9302a relayoutBoundary=up6
creator: DecoratedBox ← Container ← Listener ← _GestureSemantics ← RawGestureDetector ←
GestureDetector ← _CupertinoSegmentControlItem ← Row ← Padding ← Column ← CupertinoSegmentControl
← MediaQuery ← ⋯
decoration: BoxDecoration
RenderPadding#09d88 relayoutBoundary=up7 NEEDS-PAINT
RenderParagraph#85b48 relayoutBoundary=up8 NEEDS-PAINT
```

## Flutter Doctor
```
[✓] Flutter (on Mac OS X 10.13 17A405, locale en-GB, channel master)
    • Flutter at /Users/theobouwman/Development/flutter
    • Framework revision ad72c78342 (19 hours ago), 2017-10-16 10:58:39 -0700
    • Engine revision 7c4142808c
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.1)
    • Android SDK at /Users/theobouwman/Library/Android/sdk
    • Platform android-26, build-tools 26.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.0, Build version 9A235
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.2.5)
    • Flutter plugin version 17.0
    • Dart plugin version 172.4155.35

[✓] Connected devices
    • iPhone X • 6EE42D9C-C0E5-4E0B-AB0C-AC343B53C4EF • ios • iOS 11.0 (simulator)
```

### LINKED ISSUE #109954: [Proposal]  Revamp `Border` / Add Multiple Border Width
Right now, `Border` is inconsistent, hard to use, and different from all the other borders. I found a way to improve it, but it is not a 100% win/win scenario (hence, the proposal).

The ideas I've had were: add `BorderSide.only` (most shapes except path-based like Star would work, no breaking changes). Hixie didn't like the idea because of naming, `side` is supposed to be one side, not multiple sides, what should have multiple borders is the `Border`). So, my new proposal is refactoring `Border`. [This is the design document I wrote](https://docs.google.com/document/d/1rEQhlPzUj1QPEGSiA7eLl6kEqlkJFUecpKn-9Hcsddc/edit?usp=sharing&resourcekey=0-fpMavpYbDkqrkW2iNclV_Q) underlying the importance and possibilities. 

image

I have an implementation done for both ideas (https://github.com/bernaferrari/flutter/pull/5). Right now, I'm more focused on the Border idea, as I think it has a better chance of landing in Flutter. Border API is inconsistent:
```dart
// Currently:
Border.all(width: 3, color: Colors.white)
Border(left: BorderSide(width: 3, color: Colors.white))

// Proposal:
Border.all(width: 3, color: Colors.white)
Border(leftWidth: 3, color: Colors.white)

// Alternative proposal (.only is not currently used):
Border.only(leftWidth: 3, color: Colors.white)

```
Besides, there are multiple ways it can throw errors, which is definitely not user friendly:
```dart
// StrokeAlign must be the same.
Border(left: BorderSide(width: 5, strokeAlign: 1), right: BorderSide(width: 5, strokeAlign: 2))

// BorderRadius must be null.
BoxDecoration(radius: BorderRadius.circular(16), border: Border())

// Shape must be rectangle.
BoxDecoration(shape: BoxShape.circle, border: Border())
```

There are two uncertainties:
- Is it worth removing the option for individual colors? I believe the answer is yes.
- Should Border also have a `borderRadius` parameter? I think so, overriding the BoxDecoration if both are given. Having its `borderRadius` would make it useful in ShapeDecoration and lerp.

While the Border refactor requires deprecating half the Border class, it is not *breaking*. We can have `@Deprecated left` while we add `double leftWidth`. For simplicity, in my draft, I named the new Border as `BorderProposal` (https://github.com/bernaferrari/flutter/pull/5).

-- 
cc: hey @gspencergoog, would you like to help me with this feature? It is kind of riskier than the others, but ~you are the only one I know that likes these things~ I think we might be able to do something nice that improves the framework (again). The implementation is mostly done, depending on the path we take, the biggest challenge is convincing ourselves and the others this is worth doing.

### PR DESCRIPTION
**Pull request by bernaferrari**
Fix https://github.com/flutter/flutter/issues/12583.
Fix https://github.com/flutter/flutter/issues/109954.

Exactly like Figma:

image

image

What if it has multiple colors + borderRadius? It will crash as usual.
Side effects: material data tables will paint faster, because drrect > path.

### CODE REVIEW

File: packages/flutter/lib/src/painting/box_border.dart [hunk 1]
```diff
@@ -567,7 +632,7 @@ class Border extends BoxBorder {
     assert(() {
       if (!_strokeAlignIsUniform || top.strokeAlign != BorderSide.strokeAlignInside) {
         throw FlutterError.fromParts(<DiagnosticsNode>[
-          ErrorSummary('A Border can only draw strokeAlign different than BorderSide.strokeAlignInside on uniform borders.'),
+          ErrorSummary('A Border can only draw strokeAlign different than BorderSide.strokeAlignInside on borders with uniform colors.'),
```
Comment (bernaferrari) [inline packages/flutter/lib/src/painting/box_border.dart:635]: `uniform colors/styles` instead? Uniform colors is way more common and `style == BorderStyle.none` is kind of an invisible color.

**Comment by bernaferrari** [comment]
Hey @gspencergoog, did you miss me? 😛

File: packages/flutter/lib/src/painting/box_border.dart [hunk 2]
```diff
@@ -538,15 +578,25 @@ class Border extends BoxBorder {
       }
     }
 
+    // Allow painting non-uniform borders if the color is uniform.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/painting/box_border.dart:581]:
```suggestion
    // Allow painting non-uniform borders if the shape is rectangular, and the color and style are uniform.
```

File: packages/flutter/lib/src/painting/box_border.dart [hunk 3]
```diff
@@ -538,15 +578,25 @@ class Border extends BoxBorder {
...
+    //
+    // See also:
+    //  * <https://pub.dev/packages/non_uniform_border>, a package that
```
Comment (gspencergoog) [inline packages/flutter/lib/src/painting/box_border.dart:584]: Can you move this "See also" to a real doc comment so people can see it?

File: packages/flutter/test/painting/border_test.dart [hunk 4]
```diff
@@ -299,5 +299,21 @@ void main() {
...
+    );
+    expect(nonUniformBorder.dimensions, const EdgeInsets.fromLTRB(5, 5, 0, 20));
+
+    const BorderDirectional nonUniformBorderDirectional = BorderDirectional(
+      start: BorderSide(width: 5),
+      top: BorderSide(width: 10, strokeAlign: BorderSide.strokeAlignCenter),
+      end: BorderSide(width: 15, strokeAlign: BorderSide.strokeAlignOutside),
+      bottom: BorderSide(width: 20),
+    );
+    expect(nonUniformBorderDirectional.dimensions, const EdgeInsetsDirectional.fromSTEB(5, 5, 0, 20));
```
Comment (gspencergoog) [inline packages/flutter/test/painting/border_test.dart:317]: Can you also test to make sure that it asserts when non-uniform borders of various types are tried? You don't have to test for the assert message itself, but at least that it asserts under the right conditions.

File: packages/flutter/lib/src/painting/box_border.dart (hunk 1 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart:620]: Thanks!

File: packages/flutter/lib/src/painting/box_border.dart (hunk 2 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart:581]: I think I'll allow BoxShape.circle. I've been trying to deprecate it for 8 months, but that will give a reason for it to exist.

Let me see if I can implement non-uniform border for circles.

File: packages/flutter/test/painting/border_test.dart (hunk 4 above)
Comment (bernaferrari) [reply packages/flutter/test/painting/border_test.dart:317]: Yessss. Hold on.

**Comment by bernaferrari** [comment]
Do you think `BorderStyle.none` could be considered as `width == 0` (right now they are `color == transparent`)? If I make it the same as `width == 0` in non-uniform borders, it is one less exception. Only uniform colors would need to be required.

**Comment by gspencergoog** [comment]
>  Only uniform colors would need to be required.

Again, what happens if we add a new style?

**Comment by bernaferrari** [comment]
> Again, what happens if we add a new style?

Based on this question, I changed half my implementation to use switch so it is more future-proof 😛.

Ok, so I refactored eeeeeverything (hopefully for the better). I added support for non-uniform circle (it is a shame we can't have ovals) and improved the documentation. I got unsure if you meant the "see also" from Border or from `paint`, so I put on both (feel free to correct me which one you meant).

File: packages/flutter/lib/src/painting/box_border.dart (hunk 3 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart]: Done, check if you prefer in Border or paint (probably paint, right?).

File: packages/flutter/lib/src/painting/box_border.dart [hunk 5]
```diff
@@ -212,7 +212,8 @@ abstract class BoxBorder extends ShapeBorder {
   ///
   /// See also:
   ///
-  ///  * [paintBorder], which is used if the border is not uniform.
+  ///  * [paintBorder], which is used if the border has non-uniform colors or styles and no borderRadius.
+  ///  * [Border.paint], which is the similar method in [Border] with additional comments.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/painting/box_border.dart:216]: What do you mean by "additional comments"? More context? I'd reword this in any case.

File: packages/flutter/lib/src/painting/box_border.dart [hunk 6]
```diff
@@ -240,6 +241,64 @@ abstract class BoxBorder extends ShapeBorder {
...
+  }) {
+    final RRect borderRect;
+    switch(shape) {
+      case BoxShape.rectangle:
+        borderRect = (borderRadius ?? BorderRadius.zero)
+            .resolve(textDirection)
+            .toRRect(rect);
+        break;
+      case BoxShape.circle:
+        assert(borderRadius == null, 'A borderRadius can only be given for rectangular boxes.');
```
Comment (gspencergoog) [inline packages/flutter/lib/src/painting/box_border.dart:263]:
```suggestion
        assert(borderRadius == null, 'A borderRadius cannot be given when shape is a BoxShape.circle.');
```

File: packages/flutter/test/painting/border_test.dart [hunk 7]
```diff
@@ -299,5 +298,108 @@ void main() {
...
+        border: const Border(
+          left: BorderSide(width: 5, style: BorderStyle.none),
+          top: BorderSide(width: 10),
+          right: BorderSide(width: 15),
+          bottom: BorderSide(width: 20),
+        ),
+        borderRadius: BorderRadius.circular(25),
+      ),
+    );
+    expect(tester.takeException(), isNotNull,
```
Comment (gspencergoog) [inline packages/flutter/test/painting/border_test.dart:367]: I'd use `isAssertionError` instead of `isNotNull` to be more specific. Here and below.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/test/painting/border_test.dart (hunk 7 above)
Comment (bernaferrari) [reply packages/flutter/test/painting/border_test.dart:367]: Thanks. I didn't know about this.

File: packages/flutter/lib/src/painting/box_border.dart (hunk 5 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart:216]: Yes. I'll reword. BorderDirectional is shorter and gives less detail on what each parameter do than the equivalent Border.

Comment (bernaferrari) [reply packages/flutter/lib/src/painting/box_border.dart]: Done. Replacement is:

```dart
[Border.paint], similar to this method, includes additional comments and provides more detailed information on each parameter than described here.
```

**Comment by bernaferrari** [comment]
This will have a small, probably wanted, effect. If there is borderRadius + Border.all, the border will not be clipped around anymore (the image below). Will work like RoundedRectangleBorder.

image

**Review by Piinks — APPROVED** [review]
LGTM
